File: packages/typia/src/transformers/features/RandomTransformer.ts
import ts from "typescript";

import { RandomProgrammer } from "../../programmers/RandomProgrammer";

import { ITransformProps } from "../ITransformProps";
import { TransformerError } from "../TransformerError";

export namespace RandomTransformer {
  export const transform = (props: ITransformProps): ts.Expression => {
    // CHECK GENERIC ARGUMENT EXISTENCE
    if (!props.expression.typeArguments?.[0])
      throw new TransformerError({
        code: `typia.${props.modulo.getText()}`,
        message: "generic argument is not specified.",
      });

    // GET TYPE INFO
    const node: ts.TypeNode = props.expression.typeArguments[0];
    const type: ts.Type = props.context.checker.getTypeFromTypeNode(node);

    if (type.isTypeParameter())
      throw new TransformerError({
        code: `typia.${props.modulo.getText()}`,
        message: "non-specified generic argument.",
      });

    return ts.factory.createCallExpression(
      RandomProgrammer.write({
        context: props.context,
        modulo: props.modulo,
        type,
        name: node.getFullText().trim(),
        init: undefined,
      }),
      undefined,
      props.expression.arguments.length
        ? [props.expression.arguments[0]!]
        : undefined,
    );
  };
}

File: packages/typia/native/transform/features/RandomTransformer.go
package features

import (
  "strings"

  shimast "github.com/microsoft/typescript-go/shim/ast"
  shimscanner "github.com/microsoft/typescript-go/shim/scanner"
  nativecontext "github.com/samchon/typia/packages/typia/native/core/context"
  nativeprogrammers "github.com/samchon/typia/packages/typia/native/core/programmers"
  nativehelpers "github.com/samchon/typia/packages/typia/native/core/programmers/helpers"
  nativetransform "github.com/samchon/typia/packages/typia/native/transform/internal"
)

type randomTransformerNamespace struct{}

var RandomTransformer = randomTransformerNamespace{}

var randomTransformer_factory = shimast.NewNodeFactory(shimast.NodeFactoryHooks{})

func (randomTransformerNamespace) Transform(props nativetransform.ITransformProps) *shimast.Node {
  f := nativecontext.EmitFactoryOf(randomTransformer_factory, props.Context.Emit)
  if props.Expression == nil || props.Expression.TypeArguments == nil || len(props.Expression.TypeArguments.Nodes) == 0 {
    panic(nativetransform.NewTransformerError(nativetransform.TransformerError_IProps{
      Code:    nativehelpers.ModuloMethodText(props.Modulo),
      Message: "generic argument is not specified.",
    }))
  }
  node := props.Expression.TypeArguments.Nodes[0]
  typ := props.Context.Checker.GetTypeFromTypeNode(node)
  if typ != nil && typ.IsTypeParameter() {
    panic(nativetransform.NewTransformerError(nativetransform.TransformerError_IProps{
      Code:    nativehelpers.ModuloMethodText(props.Modulo),
      Message: "non-specified generic argument.",
    }))
  }
  name := strings.TrimSpace(shimscanner.GetTextOfNode(node))
  var arguments *shimast.NodeList
  if props.Expression.Arguments != nil && len(props.Expression.Arguments.Nodes) != 0 {
    arguments = f.NewNodeList([]*shimast.Node{props.Expression.Arguments.Nodes[0]})
  }
  return f.NewCallExpression(
    nativeprogrammers.RandomProgrammer.Write(nativeprogrammers.RandomProgrammer_IProps{
      Context: props.Context,
      Modulo:  props.Modulo,
      Type:    typ,
      Name:    &name,
      Init:    nil,
    }),
    nil,
    nil,
    arguments,
    shimast.NodeFlagsNone,
  )
}
